Feature/testcafe

ronaldvandenbroek opened the pull request:
(Pull-request not ready for merging, feature not finished. Created to track changes made in the branch in comparision to develop.)

vanhoofmaarten commented:
Hey, @ronaldvandenbroek. I've turned this PR in to a draft, because it's not ready for review yet. When ready, just click the ready for review button!

vanhoofmaarten commented:
Hi @ronaldvandenbroek, I see you've added the e2e tests script to the pre-commit hook.

Do take in consideration that the actions that we want to perform in the pre-commit hook have to be very fast to ensure a good DX (Developer Experience). For instance, when using a GIT GUI, like the one in VSCode, you're not getting a lot of visual feedback. New developers might end up wondering why their code isn't committing correctly.

At this time in the project, the tests run fast. But as time and your project progresses, test will take longer and so will running each commit.

Protecting key-branches by forcing PR's with a CI/CD workflow are the ideal middle ground of ensuring code-quality of incoming code.

vanhoofmaarten commented on tests/unit/report.spec.ts:
```diff
@@ -0,0 +1,15 @@
+const normalizeNewline = require('normalize-newline');
+const read = require('read-file-relative').readSync;
```
Have you already looked in to using the Node native 'fs' and 'path' modules? These native modules have type definitions in place. 

I don't see that much of a benefit for this particular use-case for using this dependency. If you'd really want to minimize code, you could create a separate (typed) utility function for this particular function.

```ts
import fs from 'fs';
import path from 'path';
// ...
let expected = fs.readFileSync(path.resolve(__dirname, '../data/report'), { encoding: 'utf8' });
```

* readFileSync: https://nodejs.org/dist/latest-v12.x/docs/api/fs.html#fs_fs_readfilesync_path_options
* __dirname: https://nodejs.org/api/modules.html#modules_dirname
* path.resolve: https://nodejs.org/api/path.html#path_path_resolve_paths

ronaldvandenbroek replied:
Removed the dependency on `read-file-relative`, moved report loading and normalizing to the utility folder.

ronaldvandenbroek commented on .eslintrc.js:
```diff
@@ -4,8 +4,13 @@ module.exports = {
-  },
+  overrides: [
+    {
+      files: ['**/e2e/**'],
```
There does not seem to be an easy way to enable the whitelisting of plugins within a single .eslintrc.js config file. However, a clean solution may be to make use of [cascading linting configs](https://eslint.org/docs/user-guide/configuring#configuration-cascading-and-hierarchy) that ensure that the jest plugin is only used within the unit test file by only loading the jest linting plugin in the config file present in the unit test folder.

The end result is a much cleaner .esconfig.js and the correct whitelisting behavior. It does, however, result in the use of multiple linting config files. [The suggestion can be seen here.](https://github.com/isaaceindhoven/testcafe-reporter-isaac/pull/4/commits/33060af595a212a7581acc927fa93e6412ee978b)

ronaldvandenbroek commented:
Fixed issues that came up during the stand-up and review:

- Converting all imports to require.
- Moving logic outside of unit-tests and implemented immutability.
- Whitelisting plugins with jest (see comment for more detail).
- TestCafé actually failing if a test fails instead of ignoring it.
- Splitting the linting, e2e and unit test jobs within the pipeline to account for their differences:
  - Linting does not need to run on three node versions.
  - E2e building step has been separated from the test to ensure the pipeline does not pass if the build fails.
- Trigger the pipeline only if a commit is connected to a pull-request. This can be expanded to only trigger for pull-requests to specific branches but this is not needed at the moment.

Lastly, TestCafé does not have any built-in browsers, docker might be a solution but it may be best to create a separate ticket for this. [TestCafé Browser Support](https://devexpress.github.io/testcafe/documentation/using-testcafe/common-concepts/browsers/browser-support.html)

vanhoofmaarten commented:
LGTM
